addonscript/script.py

```diff
@@ -0,0 +1,73 @@
+import aiohttp
+import asyncio
+import os
+
+import scriptworker.client
+
+from addonscript.api import do_upload, get_upload_status, get_signed_xpi
+from addonscript.task import build_filelist
+from addonscript.xpi import get_langpack_info
+
+
+def _craft_aiohttp_connector(context):
+    return aiohttp.TCPConnector()
+
+
+def get_default_config(base_dir=None):
+    base_dir = base_dir or os.path.dirname(os.getcwd())
+    default_config = {
+        'work_dir': os.path.join(base_dir, 'work_dir'),
+        'schema_file': os.path.join(os.path.dirname(__file__), 'data', 'addonscript_task_schema.json'),
+    }
+    return default_config
+
+
+async def sign_addon(context, locale):
+    # Retry Me
+    if 1:
+        upload_data = await do_upload(context, locale)
+        pk = upload_data['pk']
+    # Retry me, instead of sleeping
+    if 1:
+        await asyncio.sleep(60*6)  # wait 6 minutes
+        upload_status = await get_upload_status(context, locale, pk)
+        if len(upload_status['files']) != 1:
+            raise Exception("Expected 1 file")
+        if (upload_status.get('validation_results') and
+                upload_status['validation_results']['errors']):
+            raise Exception("Automated Validation produced errors")
+        if not upload_status['files'][0]['signed']:
+            raise Exception("expected signed xpi")
+    # Retry me
+    if 1:
+        destination = 'foo'  # XXX: Fixme
+        await get_signed_xpi(context, upload_status['files'][0]['download_url'],
+                             destination)
+
+
+def build_locales_context(context):
+    langpack_info = []
+    for f in build_filelist(context):
+        current_info = get_langpack_info(context, f)
+        langpack_info.append(current_info)
+    context.locales = {locale_info['locale']: {
+                            'unsigned': locale_info['unsigned'],
+                            'version': locale_info['version'],
+                            'id': locale_info['id'],
+                        }
+                       for locale_info in langpack_info}
+
+
+async def async_main(context):
+    connector = _craft_aiohttp_connector(context)
+    async with aiohttp.ClientSession(connector=connector) as session:
```

Comment:
I think this may have moved to scriptworker, but I think we need some preflight- and scope- checks.

CoT helps protect us against malicious users, but it doesn't protect us fully against privileged users shooting ourselves in the foot. We found preflight-checks very useful in bouncer, where we made sure no release aliases could point to beta products, etc.

Are there uses of addonscript that we would want to protect against? Checking to make sure the langpack matches the locale (I think we have this by using `get_langpack_info`), making sure we don't cross version streams, etc. could save us a lot of headache in the future.